File: 02.05.链表求和/02.05.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    res := &ListNode{}
    l1Cur, l2Cur := l1, l2
    resCur := res
    carry := 0
    for l1Cur != nil || l2Cur != nil || carry > 0 {
        sum := 0
        if l1Cur != nil {
            sum += l1Cur.Val
            l1Cur = l1Cur.Next
        }
        if l2Cur != nil {
            sum += l2Cur.Val
            l2Cur = l2Cur.Next
        }
        sum += carry
        
        carry = sum / 10
        sum = sum % 10
        resCur.Next = &ListNode{
            Val: sum,
        }
        resCur = resCur.Next
    }

    return res.Next
}

File: 02.05.链表求和/02.05.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* res = new ListNode(-1);
        ListNode* cur = res;
        int carry = 0;
        while(l1 != nullptr || l2 != nullptr || carry) {
            int sum = 0;
            if(l1 != nullptr) {
                sum += l1->val;
                l1 = l1->next;
            }
            if(l2 != nullptr) {
                sum += l2->val;
                l2 = l2->next;
            }
            sum += carry;

            carry = sum / 10;
            sum = sum % 10;
            cur->next = new ListNode(sum);
            cur = cur->next;
        }
        return res->next;
    }
};